File: stream/windows/shared/PCHUB.Streaming/Input/InputProtocol.cs
namespace PCHUB.Streaming.Input;

public static class InputProtocol
{
    public const byte MouseMove = 1;
    public const byte MouseDown = 2;
    public const byte MouseUp = 3;
    public const byte MouseWheel = 4;
    public const byte KeyDown = 5;
    public const byte KeyUp = 6;
    public const byte MouseRelMove = 7;
    public const byte MouseRelOn = 8;
    public const byte MouseRelOff = 9;

    public static byte[] MouseMove(float nx, float ny)
    {
        var buf = new byte[9];
        buf[0] = MouseMove;
        BitConverter.TryWriteBytes(buf.AsSpan(1, 4), nx);
        BitConverter.TryWriteBytes(buf.AsSpan(5, 4), ny);
        return buf;
    }

    public static byte[] MouseButton(byte button, bool down) =>
        [(down ? MouseDown : MouseUp), button];

    public static byte[] MouseWheel(short delta)
    {
        var buf = new byte[3];
        buf[0] = MouseWheel;
        BitConverter.TryWriteBytes(buf.AsSpan(1, 2), delta);
        return buf;
    }

    public static byte[] MouseRelMode(bool enabled) => [enabled ? MouseRelOn : MouseRelOff];

    public static byte[] MouseRelMove(short dx, short dy)
    {
        var buf = new byte[5];
        buf[0] = MouseRelMove;
        BitConverter.TryWriteBytes(buf.AsSpan(1, 2), dx);
        BitConverter.TryWriteBytes(buf.AsSpan(3, 2), dy);
        return buf;
    }

    public static byte[] Key(ushort scanCode, bool down, bool extended = false)
    {
        var buf = new byte[4];
        buf[0] = down ? KeyDown : KeyUp;
        BitConverter.TryWriteBytes(buf.AsSpan(1, 2), scanCode);
        if (extended) buf[3] = 1;
        return buf;
    }
}

File: web/src/lib/webrtc/input-protocol.ts
export const INPUT_CHANNEL_LABEL = "pchub-input";

export const OpMouseMove = 1;
export const OpMouseDown = 2;
export const OpMouseUp = 3;
export const OpMouseWheel = 4;
export const OpKeyDown = 5;
export const OpKeyUp = 6;
export const OpMouseRelMove = 7;
export const OpMouseRelOn = 8;
export const OpMouseRelOff = 9;

function writeF32(view: DataView, offset: number, value: number) {
  view.setFloat32(offset, value, true);
}

function writeI16(view: DataView, offset: number, value: number) {
  view.setInt16(offset, value, true);
}

export function mouseMove(nx: number, ny: number): Uint8Array {
  const buf = new ArrayBuffer(9);
  const view = new DataView(buf);
  view.setUint8(0, OpMouseMove);
  writeF32(view, 1, nx);
  writeF32(view, 5, ny);
  return new Uint8Array(buf);
}

export function mouseButton(button: number, down: boolean): Uint8Array {
  return new Uint8Array([down ? OpMouseDown : OpMouseUp, button]);
}

export function mouseWheel(delta: number): Uint8Array {
  const buf = new ArrayBuffer(3);
  const view = new DataView(buf);
  view.setUint8(0, OpMouseWheel);
  writeI16(view, 1, delta);
  return new Uint8Array(buf);
}

export function mouseRelMode(enabled: boolean): Uint8Array {
  return new Uint8Array([enabled ? OpMouseRelOn : OpMouseRelOff]);
}

export function mouseRelMove(dx: number, dy: number): Uint8Array {
  const buf = new ArrayBuffer(5);
  const view = new DataView(buf);
  view.setUint8(0, OpMouseRelMove);
  writeI16(view, 1, dx);
  writeI16(view, 3, dy);
  return new Uint8Array(buf);
}

export function keyEvent(scanCode: number, down: boolean, extended = false): Uint8Array {
  const buf = new ArrayBuffer(4);
  const view = new DataView(buf);
  view.setUint8(0, down ? OpKeyDown : OpKeyUp);
  view.setUint16(1, scanCode, true);
  view.setUint8(3, extended ? 1 : 0);
  return new Uint8Array(buf);
}

export function sendInput(channel: RTCDataChannel | null, packet: Uint8Array) {
  if (!channel || channel.readyState !== "open") return;
  try {
    const copy = new Uint8Array(packet.byteLength);
    copy.set(packet);
    channel.send(copy.buffer);
  } catch {
    // channel closed
  }
}
